### ISSUE TITLE
"Any suggestion for a good name?"

### ISSUE THREAD

**Issue by berge2007** [daniele77/cli#48]
icli

**Comment by vrobles-gee** [comment]
cli++

**Comment by daniele77** [comment]
clinteractive?

**Comment by dev-plvlml** [comment]
Maybe we should not stick to "cli" at all.

"CLI" abbreviation is most often associated with shell command-line argument parsing. E.g., practically everything mentioned in [this awesome list](https://github.com/fffaraz/awesome-cpp#cli) under the "CLI" paragraph parses command-line arguments. Specifically, the [CLI11](https://github.com/CLIUtils/CLI11) repo, already having 1.2K+ stars (and IMHO being the most convenient for the task), includes a `"CLI/CLI.hpp"` header and occupies a `CLI::` namespace, all of which is way too similar.

What this library really does (and IMHO does in the most convenient way too) is "REPL", a read-evaluate-print loop, so I suggest renaming it accordingly. For me, its strongest point is the ease of use - the fastest quickstart ever and deliberation from all the boilerplate code - just add some typed callbacks via `Insert`, and you're good to go.

Sadly, some suitable names are already taken too:
easyrepl (by a NodeJS package)
tinyrepl (by a Perl package)

Maybe we should search for a more unique word or phase? like
plug-and-repl

I'll come back with more ideas later, if you find this direction OK.

**Comment by daniele77** [comment]
@dev-plvlml, first of all, thank you very much for your contribution. 

You're right: in the programming environment, "CLI" is used mainly for parsing arguments while REPL is more appropriate. I gave the library the name "CLI" just because it's somewhat similar to CISCO CLI (actually a REPL): the interactive interface you can use to configure CISCO network nodes.
I've always been aware of this ambiguity, so much so that I often use "Interactive CLI" to distinguish it from command line parsing libraries.

Nevertheless, I think we should change it. So yes: I agree: we can investigate names derived from REPL. (REPLicate? REPLy? REPLpp? ).

I take this opportunity to summarize the naming problem.

We have two orders of problems to solve here, and I believe we should keep them apart:

- the problem of the library name (e.g., a good name should be evocative, but also easy to find on google, etc)
- the problem of the name collision in the code.

The library name should be unique, evocative, and easy to find on google. It can be descriptive (e.g., `cli`, `repl`, and so on) or it can be completely made up and meaningless (e.g., `wallaroo`, `echidna`, and so forth).
While this problem requires a careful choice of the GitHub project name, the second could be solved with the namespace name (the two could be different). 

We have these names in the library:

1. the [github] project name (the same as the root folder name) (currently `cli`)
2. the include directory (currently `include/cli`)
3. the namespace name (currently `cli`)
4. the "main" header file (currently `"cli.h"`)
5. the "main" class name (currently `Cli`)

I think that:

1. should be changed.
2. should be changed according to the project name
3. the namespace can be:
* the project name
* a different name. I would consider a name somewhat related to the author (in my case something as "d77": it's not meaningful, but it's effective for avoiding name collision!)
4. might be changed (see point 5)
5. I would not change the name of the main class, at least in the first step. This to ensure backward compatibility, to some degree (we also have `CliSession`, `CliLocalSession`, and so on...)

I would like to have a comment from you on that.

**Comment by dev-plvlml** [comment]
I agree on all the points. As to the namespace, I would prefer the project name, mainly because I can't recall any well known libraries that use a different name, so it would be counter-intuitive.

`replicate` looks like it's going to manage some prototyping. Maybe a better version for the library name would be `replicant` with nice robotic connotation.

`reply` gets closer, but instantly reminds of something remote like an HTTP server. `replay` sounds a bit better to me, like a command to a local device. But maybe this nice little thing is a `replie`?

`replpp` is okay, but there is already a [replxx](https://github.com/AmokHuginnsson/replxx), and being c++ is no more a killer feature.

I'll throw in some new suggestions:
`revaloo` - though it's hard to get that 'right' and evocable spelling.
`shimpl` - maybe it's a simple library for Shell implementation?

**Comment by ptrdtznr** [comment]
What about *clint*? *CLIINTeractive*? ;)

**Comment by SirfGuru** [comment]
i think people usually search for "c++ custom shell library" when searching for this type of library. 
chell?  (custom shell, cpp shell)
